src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.cs

```diff
@@ -30,39 +30,48 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
                         ? new CompilationData((CSharpCompilation)compilation)
                         : null);
 
-            IncrementalValuesProvider<BinderInvocation?> inputCalls = context.SyntaxProvider
+            IncrementalValueProvider<(SourceGenerationSpec?, ImmutableEquatableArray<DiagnosticInfo>?)> genSpec = context.SyntaxProvider
```

Comment:
note: diagnostics in incremental SGs are hard (especially as basically any edit invalidates them).  consider moving diagnostics, if possible, to just an analyzer that runs independently.  The analyzer can, of course, share code with the SG, esp. in the form of helper functions.

Reply:
If the diagnostics are part of the equatable incremental model, shouldn't that suffice to keep them up-to-date? IOW if an edit did trigger invalidation, then the set of diagnostics would differ from that of the cached incremental value, triggering an update to the new set downstream.

There's obviously [this issue](https://github.com/dotnet/roslyn/issues/68291) to be mindful of, which is why we're using the `DiagnosticInfo` record here instead.

Reply:
> IOW if an edit did trigger invalidation, then the set of diagnostics would differ from that of the cached incremental value, triggering an update to the new set downstream.

This is my experience as well, so I'll leave as-is in this PR.

Having a separate analyzer in the future seems like a nice idea if we can develop a clean, shared parsing abstraction with the generator implementation.

Reply:
Resolved in a few threads.